pkg/fqdn/service/service.go

```diff
@@ -102,17 +123,112 @@ type PolicyUpdater interface {
+			errChan <- err
+			return
+		}
+	}()
+	// Wait for the ACK channel to be closed
+	select {
+	case <-ackRecvStream:
+		// ACK channel is closed, we can proceed
+	case err := <-errChan:
+		// Error receiving the ACK
+		if err != nil {
+			s.log.Error("Error receiving DNS policies ACK", logfields.Error, err)
+			return err
+		}
+	}
+	s.log.Debug("Received ACK for DNS policies")
+
+	// Send the current state of the DNS policies to the client
+	s.log.Debug("Sending current state of DNS policies to the client")
+	if err := s.UpdatePolicyRules(nil, false); err != nil {
+		s.log.Error("Error sending current state of DNS rules", logfields.Error, err)
+		cancel() // Cancel the context to close the stream
+		s.deleteStream(stream)
+		return err
+	}
+
+	s.log.Debug("Waiting for context to be done for streaming DNS policies")
+	<-streamCtx.Done()
+	err := streamCtx.Err()
+	s.log.Warn("Context done for streaming DNS policies", logfields.Error, err)
+	s.deleteStream(stream)
+	return err
+}
+
+// receiveDNSPolicyACKs receives the DNS policies ACKs from the client
+// If the success is false, we can send cancel signal to the channel
+// in that case SDP will recreate the stream.
+// The flow of the method is as follows:
+// 1. Close the ACK channel for the stream to send the current state of the DNS policies
+// 2. Start a loop to receive the DNS policies ACKs
+// 3. If the context is done, return the error
+// 4. If the error is not nil, return the error
+// 5. If the response code is not NO_ERROR, send cancel signal to the channel
+// 6. Delete the request from the map
+func (s *FQDNDataServer) receiveDNSPolicyACKs(stream pb.FQDNData_StreamPolicyStateServer, ack chan struct{}) error {
+	var once sync.Once
+
+	for {
+		select {
+		case <-stream.Context().Done():
+			s.log.Info("Stream context is finished, closing the stream")
+			return stream.Context().Err()
+		default:
+			// Continue receiving the DNS policies ACKs
+		}
+		once.Do(func() {
+			close(ack)
+		})
+		response, err := stream.Recv()
+		if err != nil {
+			return err
+		}
+		s.log.Debug("Received DNS policies ACK", logfields.Response, response)
+		requestId := response.GetRequestId()
+		_, ok := s.policyRulesResponse.Load(requestId)
+		if !ok {
+			s.log.Warn("Received Message id not found", logfields.ID, requestId)
+		} else {
+			s.log.Debug("Received response for dns message id", logfields.ID, requestId)
+
+			// We can send cancel signal to the channel if the response code is not NO_ERROR,
+			// in that case SDP will recreate the stream.
+			responseCode := response.GetResponse()
+			if responseCode != pb.ResponseCode_RESPONSE_CODE_NO_ERROR {
+				s.log.Error("Received error response for dns message id", logfields.ID, requestId)
+				cancel, ok := s.streams.Load(stream)
+				if ok {
+					cancel()
+				}
+			}
+
+			// Delete the request from the map
+			s.policyRulesResponse.Delete(requestId)
+			s.log.Debug("Deleted request id from the map: ", logfields.ID, requestId)
```

Comment:
Don't end log messages with ": ". There is no guarantee in what order the log attributes are shown, so likely the "ID" won't follow the ": " suffix. This applies to multiple places in this file (maybe others too?).